samza-core/src/main/java/org/apache/samza/table/caching/CachingTable.java

```diff
@@ -0,0 +1,152 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.samza.table.caching;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+import org.apache.samza.storage.kv.Entry;
+import org.apache.samza.table.ReadWriteTable;
+import org.apache.samza.table.ReadableTable;
+
+import com.google.common.base.Preconditions;
+
+
+/**
+ * A composite table incorporating a cache with a Samza table. The cache is
+ * represented as a {@link ReadWriteTable}. Optionally if the table is writable,
+ * CachingTable supports both write-through and write-around policy.
+ *
+ * @param <K> type of the table key
+ * @param <V> type of the table value
+ */
+public class CachingTable<K, V> implements ReadWriteTable<K, V> {
+  private final ReadableTable<K, V> rdTable;
+  private final ReadWriteTable<K, V> rwTable;
+  private final ReadWriteTable<K, V> cache;
+  private final boolean isWriteAround;
+
+  /**
+   * This is used for fine-grained locking for concurrent accesses
+   * at key-level to allow more parallelism for get/put operations
+   * among distinct keys.
+   *
+   * We leverage {@link ConcurrentHashMap#computeIfAbsent(Object, java.util.function.Function)}.
+   * for per-key atomic operation. For batch modifications (putAll/deleteAll),
+   * we synchronize the whole table instead.
+   */
+  private final ConcurrentHashMap<K, V> locker;
+
+  public CachingTable(ReadableTable<K, V> table, ReadWriteTable<K, V> cache, boolean isWriteAround) {
+    this.rdTable = table;
+    this.rwTable = table instanceof ReadWriteTable ? (ReadWriteTable) table : null;
+    this.cache = cache;
+    this.isWriteAround = isWriteAround;
+    this.locker = new ConcurrentHashMap();
+  }
+
+  @Override
+  public V get(K key) {
+    V value = cache.get(key);
+    if (value != null) {
+      return value;
+    }
+
+    // Atomically populates the value for a cache miss
+    locker.computeIfAbsent(key, (k) -> {
```

Comment:
locker never holds the data, hence the "return null" in the end. It's only used for atomically update the cache/table for a key, once it's done, it's never entered into the locker. To put it simply, we use the concurrent hash map (locker) not a map but as a key-based synchronization tool.

Reply:
Consider this case: when you enter the lock section, the key is still in the cache, so you do nothing there. right after that the key is evicted from the cache. Would you not end up returning null then?

Reply:
if key exists, get() would have returned already at line 71. If it's added after 71, it means it's a new key, assuming the TTL is not at usec level, I don't see how a key gets evited in the middle of the locker right?

Reply:
Sorry to barge in..

If the CHM is never written to then it will never expand, and therefore have the default number of locks (16). You would be better served using Guava's `Striped` instead, if that's all you need.

Reply:
@ben-manes Appreciate the excellent suggestion! CHM.putIfAbsent is so convenient so I went with that. But having the flexibility for setting the stripes can be useful for tweaking the parallelism. I'll read up more on Striped and update the PR.

Reply:
Ideally in Guava you would delegate the loading to it (via `LoadingCache` or `get(key, callable)`. I'm guessing you didn't do that to avoid changing the existing interfaces, but that would be ideal otherwise.

A self-serving FYI, but you might also consider [Caffeine](https://github.com/ben-manes/caffeine) over Guava, which is a modernized rewrite of their cache.

Reply:
@ben-manes Yeah, I've read about caffeine and the excellent performance. I think we can certainly implement a Caffeine cache table down the line given right now Samza does not have a dependency on Caffeine. 

Again, really appreciate the inputs!